src/v/utils/xid.cc

```diff
@@ -0,0 +1,133 @@
+  'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v'};
+
+using decoder_t = std::array<uint8_t, 256>;
+
+constexpr decoder_t build_decoder_table() {
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+    std::array<uint8_t, 256> ret = {0xff};
+
+    for (uint8_t i = 0; i < b32_alphabet.size(); ++i) {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-constant-array-index)
+        ret[b32_alphabet[i]] = i;
+    }
+    return ret;
+}
+
+static constexpr decoder_t decoder = build_decoder_table();
+} // namespace
+
+invalid_xid::invalid_xid(const ss::sstring& current_string)
+  : _msg(ssx::sformat("String '{}' is not a valid xid", current_string)) {}
+
+xid xid::from_string(const ss::sstring& str) {
+    if (str.size() != str_size) {
+        throw invalid_xid(str);
+    }
+    // NOLINTBEGIN(cppcoreguidelines-pro-bounds-constant-array-index,
+    // readability-magic-numbers, hicpp-signed-bitwise)
+    data_t data;
+    data[11] = decoder[str[17]] << 6u | decoder[str[18]] << 1u
+               | decoder[str[19]] >> 4u;
+
+    // check the last byte
+    if (b32_alphabet[(data[11] << 4) & 0x1F] != str[19]) {
+        throw invalid_xid(str);
+    }
+
+    data[10] = decoder[str[16]] << 3u | decoder[str[17]] >> 2u;
+    data[9] = decoder[str[14]] << 5u | decoder[str[15]];
+    data[8] = decoder[str[12]] << 7u | decoder[str[13]] << 2u
+              | decoder[str[14]] >> 3u;
+    data[7] = decoder[str[11]] << 4u | decoder[str[12]] >> 1;
+    data[6] = decoder[str[9]] << 6u | decoder[str[10]] << 1u
+              | decoder[str[11]] >> 4u;
+    data[5] = decoder[str[8]] << 3u | decoder[str[9]] >> 2u;
+    data[4] = decoder[str[6]] << 5u | decoder[str[7]];
+    data[3] = decoder[str[4]] << 7u | decoder[str[5]] << 2u
+              | decoder[str[6]] >> 3u;
+    data[2] = decoder[str[3]] << 4u | decoder[str[4]] >> 1u;
+    data[1] = decoder[str[1]] << 6u | decoder[str[2]] << 1
+              | decoder[str[3]] >> 4u;
+    data[0] = decoder[str[0]] << 3u | decoder[str[1]] >> 2u;
+    // NOLINTEND(cppcoreguidelines-pro-bounds-constant-array-index,
+    // readability-magic-numbers, hicpp-signed-bitwise)
+    return xid(data);
+}
+
+fmt::format_context::iterator
+fmt::formatter<xid>::format(const xid& id, format_context& ctx) const {
+    static constexpr uint8_t mask = 0x1f;
+    // NOLINTBEGIN(cppcoreguidelines-pro-bounds-constant-array-index,
+    // readability-magic-numbers, hicpp-signed-bitwise)
+    fmt::format_to(
+      ctx.out(),
+      "{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}",
+      b32_alphabet[id._data[0] >> 3u],
+      b32_alphabet[((id._data[1] >> 6u) & mask) | ((id._data[0] << 2u) & mask)],
+      b32_alphabet[(id._data[1] >> 1) & mask],
+      b32_alphabet[((id._data[2] >> 4) & mask) | ((id._data[1] << 4) & mask)],
+      b32_alphabet[id._data[3] >> 7 | ((id._data[2] << 1) & mask)],
+      b32_alphabet[(id._data[3] >> 2) & mask],
+      b32_alphabet[id._data[4] >> 5 | ((id._data[3] << 3) & mask)],
```

Comment:
Is this producing the same format as https://github.com/rs/xid/blob/master/id.go#L170 ?

Reply:
yes